docs/docs/dev_docs/tutorials/writing_private_voting_contract.md

```diff
@@ -0,0 +1,251 @@
+
+#include_code storage_struct yarn-project/noir-contracts/src/contracts/easy_private_voting_contract/src/main.nr rust
+
+In this contract, we will store three vars:
+1. admin, as an Aztec address held in public state
+2. tally, as a map of Fields held in public state
+3. voteEnded, as a boolean held in public state
+
+Under the struct, define the impl block like this:
+
+#include_code storage_impl yarn-project/noir-contracts/src/contracts/easy_private_voting_contract/src/main.nr rust
+
+The `impl` block must define one function `init` that explains how to access and manipulate our variables. We pass context, a storage slot, and serialization methods we imported earlier. 
+
+This `init` function will be called every time we access `storage` in our functions.
+
+## Constructor
+
+The next step is to initialize the contract with a constructor. The constructor will take an address as a parameter and set the admin.
+
+All constructors must be private, and because the admin is in public storage, we cannot directly update it from the constructor. You can find more information about this [here](../../concepts/foundation/communication/public_private_calls.md).
+
+Therefore our constructor must call a public function by using `context.call_public_function()`. Paste this under the `impl` storage block:
+
+#include_code constructor yarn-project/noir-contracts/src/contracts/easy_private_voting_contract/src/main.nr rust
+
+`context.call_public_function()` takes three arguments:
+1. The caller of the function (to keep things private, we are using `context.this_address()` which refers to the contract address)
+2. The selector of the function to call (we can use `compute_selector()` for this)
+3. The arguments of the function (we pass the `admin`)
+
+We now need to write the `_initialize()` function:
+
+#include_code initialize yarn-project/noir-contracts/src/contracts/easy_private_voting_contract/src/main.nr rust
+
+This function takes the admin argument and writes it to the storage. We are also using this function to set the `voteEnded` boolean as false in the same way.
+
+This function is set as `internal` so that it can only be called from within the contract. This stops anyone from setting a new admin.
+
+## Casting a vote privately
+
+For the sake of simplicity, we will assume three things:
+1. Everyone with an Aztec account gets a vote
+2. They can only vote once in this contract
+3. Who they are is private, but their actual vote is not
+
+Initialize a private function called `cast_vote`:
+```rust
+#include_code initialize yarn-project/noir-contracts/src/contracts/easy_private_voting_contract/src/main.nr raw
+}
+```
+
+To ensure someone only votes once, we will create a nullifier. If they try to vote again, a nullifier will be found and the function will stop.
+
+We could create a nullifier with the address directly, but this would be easy to reverse-engineer. We should instead create a nullifier with the account's secret key:
```

Comment:
```suggestion
We could create a nullifier with the address directly, but this leaks privacy as it would be easy to reverse-engineer. We must add some randomness or some form of secret, like nullifier secrets **[INSERT NULLIFIER SECRET reference from the docs!**]. We should instead create a nullifier with the account's secret key:
```